<a href="https://colab.research.google.com/github/DrSubbiah/6.Python/blob/main/3_Python_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#<font color='darkred'> Pandas Introduction

## <font color='darkblue'>What is Pandas?

Pandas is a Python library used for working with data sets.

It has functions for analyzing, cleaning, exploring, and manipulating data.

The name "Pandas" has a reference to both "Panel Data", and "Python Data Analysis" and was created by Wes McKinney in 2008.

##<font color='darkblue'> Why Use Pandas?

Pandas allows us to analyze big data and make conclusions based on statistical theories.

Pandas can clean messy data sets, and make them readable and relevant.

Relevant data is very important in data science.

In [ ]:
import pandas as pd

In [ ]:
mydataset = {
  'cars': ["BMW", "Volvo", "Ford"],
  'passings': [3, 7, 2]
}

myvar = pd.DataFrame(mydataset)

print(myvar)

# <font color='darkred'>Pandas Series

## <font color='darkblue'>What is a Series?

A Pandas Series is like a column in a table.

It is a one-dimensional array holding data of any type.

pd.read_csv('/content/Algo0_PP.csv')

In [ ]:
a = [1, 7, 2]
myvar = pd.Series(a)
print(myvar)


# <font color='darkred'> Labels
If nothing else is specified, the values are labeled with their index number. First value has index 0, second value has index 1 etc.

This label can be used to access a specified value.

Return the first value of the Series:print(myvar[0])
    
Create Labels

With the index argument, you can name your own labels.

In [ ]:
a = [1, 7, 2]
myvar = pd.Series(a)
myvar1 = pd.Series(a, index = ["x", "y", "z"])

myvar2=pd.Series(a, name='Numbers')

print(myvar)
print(myvar1)
print(myvar2)

# <font color='darkred'> Key/Value Objects as Series
You can also use a key/value object, like a dictionary, when creating a Series.

Create a simple Pandas Series from a dictionary:

In [ ]:
calories = {"day1": 420, "day2": 380, "day3": 390}
myvar = pd.Series(calories)
print(myvar)

day1    420
day2    380
day3    390
dtype: int64


#<font color='darkred'> Named Indexes in Series
With the index argument, you can name your own indexes.

Add a list of names to give each row a name:

###<font color='darkgreen'> Locate Named Indexes *loc* attribute
Use the named index in the loc attribute to return the specified row(s).

Return "day2":

In [ ]:
myvar.loc['day2']

380

#<font color='darkred'> DataFrames

Data sets in Pandas are usually multi-dimensional tables, called DataFrames.

A Pandas DataFrame is a 2 dimensional data structure, like a 2 dimensional array, or a table with rows and columns.

Series is like a column, a DataFrame is the whole table.

Create a DataFrame from two Series:

In [ ]:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}

#load data into a DataFrame object:
df = pd.DataFrame(data)

print(df)

## <font color='darkblue'>List to DF

In [ ]:
# list of strings
lst = ['alpha', 'beta', 'gamma', 'delta', '0', '1', '2']

# Calling DataFrame constructor on list
df = pd.DataFrame(lst)
print(df)

# <font color='darkred'>Read / Load a CSV

In [ ]:
from google.colab import files
uploaded = files.upload()

In [ ]:
df = pd.read_csv('/content/Algo0_PP.csv')

# <font color='darkred'>DATA INFORMATION
The DataFrames object has a method called info(), that gives you more information about the data set.

In [ ]:
print(df)

In [ ]:
print(df.info())

In [ ]:
print(df.dtypes)

In [ ]:
print(df.shape)
print(df.ndim)
print(df.size)

In [ ]:
print(df.head(5))

In [ ]:
print(df.tail(5))

In [ ]:
print(df.columns)    # for reading the headers

#<font color='darkred'> DATA HANDLING

##<font color='darkblue'> Subsetting and Rransformations

### <font color='darkgreen'> Column Name and / or Indices

In [ ]:
print(df.symbol) #Extract the column "symbol"

In [ ]:
print(df['symbol'][0:6])    # give the column name for printing all the elements from the chosen rows

0          SBIN
1    AUROPHARMA
2          ISEC
3      CHOLAFIN
4      HINDALCO
5    TATAMOTORS
Name: symbol, dtype: object


In [ ]:
print(df[['symbol', 'date', 'time']])  # giving multiple column names inside lists - two square brackets

## <font color='darkblue'> Locate Row

A DataFrame is a 2-dimensional table with rows and columns.

Pandas use the **loc** attribute to return one or more specified row(s)

In [ ]:
print(df.loc[0]) #Return row 0

print(df.loc[[0, 1]]) #Return rows 0 and 1

print(df.loc[0:5]) #Return rows 0 to 5 (both inclusive)

## <font color='darkblue'> *loc* and column names

In [ ]:
df.loc[0:5]['symbol']

df.loc[0:5][['symbol','pp']] #Multiple columns

## <font color='darkblue'> usage of *iloc*

In [ ]:
print(df.iloc[2:6])  # print elements from specified range
print(df.loc[2:6]) #Both are same

In [ ]:
print(df.iloc[2, 1])    # Extract 3rd row and 2nd column - passing index (rows) and column numbers as arguments
print(df.loc[2])        # Extract third row

In [ ]:
print(df.iloc[2:6, 1]) # Extract the required index / column

print(df.iloc[2:6,1:3])

## <font color='darkblue'>Selected Rows, all Columns

## <font color='darkblue'>Selected Columns, all Rows

In [ ]:
print(df.iloc[2:6,:])

print(df.iloc[ :,0:2])

In [ ]:
for index, row in df.iterrows():            #Iterate over DataFrame rows
  print(index, row['symbol'])   # or just put print(index, row)

## <font color='darkblue'> Filter (subset) using *loc*

In [ ]:
df.loc[df['symbol']=='SBIN']

In [ ]:
df.loc[(df['symbol']=='SBIN')| (df['symbol']=='LUPIN')] # Usage of or logic - symbol "|"

In [ ]:
df.loc[(df['symbol']=='SBIN')& (df['cycle']%28!=0)] # Usage of and logic - symbol "&"

In [ ]:
df.loc[(df['symbol']=='SBIN')& (df['cycle']%28==0)]

In [ ]:
df.loc[((df['symbol']=='SBIN')|(df['symbol']=='LUPIN')) & (df['cycle']%28==0)]

## <font color='darkblue'> *loc* within a column

Filtering can be done on a single or multiple columns conditioned on other columns

In [ ]:
ex1=df.loc[(df['cycle']%28==0)&(df['cycle']>1)&(df['pp']<1000)] #Whole data set

ex2=df['symbol'].loc[(df['cycle']%28==0)&(df['cycle']>1)&(df['pp']<1000)] #Only one column

print(ex1)
print(ex2)
print(ex2.unique())

      table_id      symbol        date      time   pp   pr   ps  cycle
1097     15719  AUROPHARMA  19-01-2023  09.43.02  0.0  0.0  0.0     28
1098     15720    CHOLAFIN  19-01-2023  09.43.02  0.0  0.0  0.0     28
1101     15723  TATACONSUM  19-01-2023  09.43.02  0.0  0.0  0.0     28
1102     15724    HINDALCO  19-01-2023  09.43.02  0.0  0.0  0.0     28
1103     15725        SBIN  19-01-2023  09.43.02  0.0  0.0  0.0     28
...        ...         ...         ...       ...  ...  ...  ...    ...
5490     20112    HDFCLIFE  19-01-2023  11.37.02  0.0  0.0  0.0    140
5510     20132  ADANIPORTS  19-01-2023  11.37.03  0.0  0.0  0.0    140
5526     20148         TCS  19-01-2023  11.37.04  0.0  0.0  0.0    140
5527     20149        VEDL  19-01-2023  11.37.04  0.0  0.0  0.0    140
5556     20178       LUPIN  19-01-2023  11.38.03  0.0  0.0  0.0    140

[191 rows x 8 columns]
1097    AUROPHARMA
1098      CHOLAFIN
1101    TATACONSUM
1102      HINDALCO
1103          SBIN
           ...    
5490      

In [ ]:
df[['symbol','pr']].loc[(df['pp']>1000)&(df['cycle']>1)]

In [ ]:
df.loc[df['cycle']%28==0]

## <font color='darkblue'>Sort a DF based on a column

In [ ]:
df.sort_values('symbol', ascending = False) #default: True

## <font color='darkblue'>Creating a new column

###<font color='darkgreen'> Addition of existing two numeric columns

In [ ]:
df['Total'] = df['pr'] + df['pp']
print(df['Total'])
#Alternately, use column index
df1=df.iloc[:, 4]+df.iloc[:, 5]
print(df1)
#or
df2=df['Total'] = df.iloc[:, 4:6].sum(axis=1)
print(df1)

# <font color='darkblue'> Named Index in a DF

In [ ]:
data = {
  "calories": [420, 380, 390],
  "duration": [50, 40, 45]
}
EX = pd.DataFrame(data, index = ["day1", "day2", "day3"])
print(EX)

In [ ]:
print(EX.loc["day2"])

print(EX.loc[["day2",'day3']])

# Pandas - Cleaning Data

Data cleaning means fixing bad data in your data set.

  * Bad data could be:

  * Empty cells

  * Data in wrong format

  * Wrong data

  * Duplicates

# Pandas - Cleaning Empty Cells

Empty cells can potentially give you a wrong result when you analyze data.

Remove Rows

One way to deal with empty cells is to remove rows that contain empty cells.

**If the data set is sufficiently large then removing a few rows may not have a big impact on the result**   

Return a new Data Frame with no empty cells:

In [ ]:
import pandas as pd
df = pd.read_csv('C://Users//lmohan2//Desktop//New folder//test//dirtydata.csv')
new_df = df.dropna()
print(new_df.to_string())

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

If you want to change the original DataFrame, use the inplace = True argument:

Remove all rows with NULL values:

In [ ]:
import pandas as pd
df = pd.read_csv('C://Users//lmohan2//Desktop//New folder//test//dirtydata.csv')
df.dropna(inplace = True)
print(df.to_string())

    Duration          Date  Pulse  Maxpulse  Calories
0         60  '2020/12/01'    110       130     409.1
1         60  '2020/12/02'    117       145     479.0
2         60  '2020/12/03'    103       135     340.0
3         45  '2020/12/04'    109       175     282.4
4         45  '2020/12/05'    117       148     406.0
5         60  '2020/12/06'    102       127     300.0
6         60  '2020/12/07'    110       136     374.0
7        450  '2020/12/08'    104       134     253.3
8         30  '2020/12/09'    109       133     195.1
9         60  '2020/12/10'     98       124     269.0
10        60  '2020/12/11'    103       147     329.3
11        60  '2020/12/12'    100       120     250.7
12        60  '2020/12/12'    100       120     250.7
13        60  '2020/12/13'    106       128     345.3
14        60  '2020/12/14'    104       132     379.3
15        60  '2020/12/15'     98       123     275.0
16        60  '2020/12/16'     98       120     215.2
17        60  '2020/12/17'  

# Replace Empty Values
Another way of dealing with empty cells is to insert a new value instead.

This way you do not have to delete entire rows just because of some empty cells.

The fillna() method allows us to replace empty cells with a value:

Replace NULL values with the number 130:

In [ ]:
import pandas as pd
df = pd.read_csv('C://Users//lmohan2//Desktop//New folder//test//dirtydata.csv')
df.fillna(130, inplace = True)

# Replace Only For a Specified Columns
The example above replaces all empty cells in the whole Data Frame.

To only replace empty values for one column, specify the column name for the DataFrame:

Replace NULL values in the "Calories" columns with the number 130:

In [ ]:
import pandas as pd
df = pd.read_csv('C://Users//lmohan2//Desktop//New folder//test//dirtydata.csv')
df["Calories"].fillna(130, inplace = True)

# Replace Using Mean, Median, or Mode
A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

Calculate the MEAN, and replace any empty values with it:

In [ ]:
import pandas as pd
df = pd.read_csv('C://Users//lmohan2//Desktop//New folder//test//dirtydata.csv')
x = df["Calories"].mean()
df["Calories"].fillna(x, inplace = True)

Calculate the MEDIAN, and replace any empty values with it:

In [ ]:
import pandas as pd
#df = pd.read_csv('data.csv')
x = df["Calories"].median()
df["Calories"].fillna(x, inplace = True)

# Pandas - Cleaning Data of Wrong Format
Data of Wrong Format
Cells with data of wrong format can make it difficult, or even impossible, to analyze data.

To fix it, you have two options: remove the rows, or convert all cells in the columns into the same format.

Convert Into a Correct Format
In our Data Frame, we have two cells with the wrong format. Check out row 22 and 26, the 'Date' column should be a string that represents a date:

Let's try to convert all cells in the 'Date' column into dates.

Pandas has a to_datetime() method for this:

In [ ]:
import pandas as pd
df = pd.read_csv('C://Users//lmohan2//Desktop//New folder//test//dirtydata.csv')
df['Date'] = pd.to_datetime(df['Date'])
print(df.to_string())

    Duration       Date  Pulse  Maxpulse  Calories
0         60 2020-12-01    110       130     409.1
1         60 2020-12-02    117       145     479.0
2         60 2020-12-03    103       135     340.0
3         45 2020-12-04    109       175     282.4
4         45 2020-12-05    117       148     406.0
5         60 2020-12-06    102       127     300.0
6         60 2020-12-07    110       136     374.0
7        450 2020-12-08    104       134     253.3
8         30 2020-12-09    109       133     195.1
9         60 2020-12-10     98       124     269.0
10        60 2020-12-11    103       147     329.3
11        60 2020-12-12    100       120     250.7
12        60 2020-12-12    100       120     250.7
13        60 2020-12-13    106       128     345.3
14        60 2020-12-14    104       132     379.3
15        60 2020-12-15     98       123     275.0
16        60 2020-12-16     98       120     215.2
17        60 2020-12-17    100       120     300.0
18        45 2020-12-18     90 

# Pandas - Fixing Wrong Data
Wrong Data
"Wrong data" does not have to be "empty cells" or "wrong format", it can just be wrong, like if someone registered "199" instead of "1.99".

Sometimes you can spot wrong data by looking at the data set, because you have an expectation of what it should be.

If you take a look at our data set, you can see that in row 7, the duration is 450, but for all the other rows the duration is between 30 and 60.

It doesn't have to be wrong, but taking in consideration that this is the data set of someone's workout sessions, we conclude with the fact that this person did not work out in 450 minutes.

How can we fix wrong values, like the one for "Duration" in row 7?

# Replacing Values
One way to fix wrong values is to replace them with something else.

In our example, it is most likely a typo, and the value should be "45" instead of "450", and we could just insert "45" in row 7:

Set "Duration" = 45 in row 7:

In [ ]:
df.loc[7, 'Duration'] = 45

For small data sets you might be able to replace the wrong data one by one, but not for big data sets.

To replace wrong data for larger data sets you can create some rules, e.g. set some boundaries for legal values, and replace any values that are outside of the boundaries.

Loop through all values in the "Duration" column.

If the value is higher than 120, set it to 120:

In [ ]:
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.loc[x, "Duration"] = 120

# Removing Rows
Another way of handling wrong data is to remove the rows that contains wrong data.

This way you do not have to find out what to replace them with, and there is a good chance you do not need them to do your analyses.

Delete rows where "Duration" is higher than 120:


In [ ]:
for x in df.index:
  if df.loc[x, "Duration"] > 120:
    df.drop(x, inplace = True)

# Pandas - Removing Duplicates
Discovering Duplicates
Duplicate rows are rows that have been registered more than one time.

By taking a look at our test data set, we can assume that row 11 and 12 are duplicates.

To discover duplicates, we can use the duplicated() method.

The duplicated() method returns a Boolean values for each row:

Returns True for every row that is a duplicate, othwerwise False:

In [ ]:
print(df.duplicated())

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12     True
13    False
14    False
15    False
16    False
17    False
18    False
19    False
20    False
21    False
22    False
23    False
24    False
25    False
26    False
27    False
28    False
29    False
30    False
31    False
dtype: bool


# Removing Duplicates
To remove duplicates, use the drop_duplicates() method.

Remove all duplicates:



In [ ]:
df.drop_duplicates(inplace = True)

# Pandas - Data Correlations
Finding Relationships

A great aspect of the Pandas module is the corr() method.

The corr() method calculates the relationship between each column in your data set.

Show the relationship between the columns:

In [ ]:
df.corr()

,Duration,Pulse,Maxpulse,Calories
Duration,1.000000,-0.059452,-0.250033,0.356437
Pulse,-0.059452,1.000000,0.269672,0.506593
Maxpulse,-0.250033,0.269672,1.000000,0.344494
Calories,0.356437,0.506593,0.344494,1.000000


# Result Explained
The Result of the corr() method is a table with a lot of numbers that represents how well the relationship is between two columns.

The number varies from -1 to 1.

1 means that there is a 1 to 1 relationship (a perfect correlation), and for this data set, each time a value went up in the first column, the other one went up as well.

0.9 is also a good relationship, and if you increase one value, the other will probably increase as well.

-0.9 would be just as good relationship as 0.9, but if you increase one value, the other will probably go down.

0.2 means NOT a good relationship, meaning that if one value goes up does not mean that the other will.

Perfect Correlation:
We can see that "Duration" and "Duration" got the number 1.000000, which makes sense, each column always has a perfect relationship with itself.

Good Correlation:
"Duration" and "Calories" got a 0.922721 correlation, which is a very good correlation, and we can predict that the longer you work out, the more calories you burn, and the other way around: if you burned a lot of calories, you probably had a long work out.

Bad Correlation:
"Duration" and "Maxpulse" got a 0.009403 correlation, which is a very bad correlation, meaning that we can not predict the max pulse by just looking at the duration of the work out, and vice versa.